MT3-439 #ready-for-test Create second failed attempt page

LBHABEBIC opened the pull request:
<!--
  Is this related to a ticket in Jira? If so, add Smart Commit commands to the
  PR title to be automatically included in the merge commit if the PR is merged.

  For example:

  MT3-221 #ready-for-test #comment Create a pull request template

  https://confluence.atlassian.com/fisheye/using-smart-commits-960155400.html
  -->

# What?

<!--
  What new features or changes does this pull request contain?

  Include before and after screenshots, if appropriate.
  -->

# Why?

<!--
  What motivates these changes?

  Include any user stories driving this feature, if appropriate.
  -->

# Notes

<!--
  Is there anything about the implementation worth calling out to reviewers?
  -->

# Next steps

<!--
  Is there any follow up work that needs to be done?

  Consider using a checklist, for example:

  - [ ] do something
  - [ ] do something else
  -->

LBHABEBIC committed:
Add Second Failed Attempt step

Displaying date of first visit

erbridge commented on steps/PageSlugs.ts:
```diff
@@ -131,9 +133,9 @@ export const repeatingStepSlugs = [
 
 export const urlObjectForSlug = (
   router: NextRouter,
-  slug: string
+  slug: string | undefined
 ): { pathname: string } => {
-  return prefixUrl(router, { pathname: `/${slug}` });
+  return prefixUrl(router, { pathname: slug === undefined ? "" : `/${slug}` });
```
`slug` should never be undefined. If it is, resolve that before calling this function.

erbridge commented on __tests__/features/accessibility.feature:
```diff
@@ -279,4 +279,11 @@ Feature: Accessibility
+    And I wait for the data to be fetched
+    And I visit /second-failed-attempt for the process
+    Then the page should be accessible
     And the page should have a descriptive title
```
Trailing new line, please.

erbridge commented on steps/unable-to-enter/second-failed-attempt.tsx:
```diff
@@ -0,0 +1,221 @@
+      await stores.unableToEnter.put(processRef, {
+        ...unableToEnter,
+        secondFailedAttempt: {
+          ...unableToEnter.secondFailedAttempt,
```
This method looks identical to the first failed attempt's version, except for the key to use. Let's refactor it into a helper.

erbridge commented on steps/unable-to-enter/second-failed-attempt.tsx:
```diff
@@ -0,0 +1,221 @@
+    .map((checkbox) => checkbox.label);
+
+  return (
+    <PageAnnouncement title={"Previous attempt"}>
```
```suggestion
    <PageAnnouncement title="Previous attempts">
```

erbridge commented on steps/unable-to-enter/second-failed-attempt.tsx:
```diff
@@ -0,0 +1,221 @@
+    <PageAnnouncement title={"Previous attempt"}>
+      <List
+        items={[
+          <React.Fragment key="first">
+            First failed attempt: {dateString}
+          </React.Fragment>,
+          <React.Fragment key="second">
+            Reason: <List items={reasonLabel.map((label) => label)} />
+          </React.Fragment>,
```
This created multiple list items for what should probably be a single item. Try:

```suggestion
          <React.Fragment key="first-attempt">
            <div>First failed attempt: {dateString}</div>
            <div>Reason: <List items={reasonLabel.map((label) => label)} /></div>
          </React.Fragment>,
```

Where the `div`s might be better as `Paragraph`s.

erbridge commented on steps/unable-to-enter/second-failed-attempt.tsx:
```diff
@@ -0,0 +1,221 @@
+                anything of concern?
+              </FieldsetLegend>
+            ) as React.ReactNode,
+            checkboxes: [
+              {
+                label: "Tenant not in",
+                value: "tenant not in",
+              },
+              {
+                label: "Not safe to enter property",
+                value: "not safe",
+              },
+              {
+                label: "Property may be abandoned",
+                value: "abandoned",
+              },
+              {
+                label: "There are signs of subletting",
+                value: "subletting",
+              },
+              {
+                label: "Other",
+                value: "other",
+              },
+            ],
```
```suggestion
            checkboxes: failedAttemptReasonCheckboxes
```

And do the same for the first failed attempt.

Gweaton committed:
Add second-failed-attempt to the journey

Users will be taken to second failed attempt page if a first attempt to
already been made

Gweaton committed:
Extract date persistence into shared function

erbridge commented on helpers/persistUnableToEnterDate.ts:
```diff
@@ -0,0 +1,44 @@
+}
+
+export const persistUnableToEnterDate = async (
+  attemptKey: FailedAttempts
```
```suggestion
  attemptKey: keyof ProcessDatabaseSchema["schema"]["unableToEnter"]["value"]
```

erbridge replied:
Oh I see. We use these elsewhere. Leaving the comment as it might be helpful anyway.

erbridge commented on steps/previsit/outside.ts:
```diff
@@ -1,3 +1,4 @@
+import { makeUnableToEnterSubmit } from "components/makeUnableToEnterSubmit";
```
```suggestion
import { makeUnableToEnterSubmit } from "../../components/makeUnableToEnterSubmit";
```

Gweaton committed:
Extract function to determine failed attempt slug

This will encapsulate the logic which will include a few more cases as
we add more steps

erbridge commented on components/SubmitButtons.tsx:
```diff
@@ -0,0 +1,93 @@
+import PropTypes from "../helpers/PropTypes";
+import urlsForRouter from "../helpers/urlsForRouter";
+import PageSlugs, { urlObjectForSlug } from "../steps/PageSlugs";
+import { MakeSubmitProps } from "./makeSubmit";
```
Not resolved.

Gweaton replied:
Ugh, sorry - moved it into the wrong place

erbridge commented on components/makeUnableToEnterSubmit.tsx:
```diff
@@ -0,0 +1,62 @@
+  > = ({ disabled, onSubmit }) => {
+    const buttonProps = [
+      props,
+      { slug: useUnableToEnterSlug(), value: "Unable to Enter" },
```
Let's pull the hook up to assign to a variable. Makes it easier to spot that there's a hook.

erbridge commented on helpers/persistUnableToEnterDate.ts:
```diff
@@ -0,0 +1,44 @@
+    return;
+  }
+
+  const processRef = keyFromSlug()();
```
```suggestion
  const processRef = getProcessRef();
```

erbridge reviewed (commented):
Nearly there.
